### ISSUE TITLE
Type error for UISchema object

### ISSUE THREAD

**Issue by dcantu96** [eclipsesource/jsonforms#2109]
### Describe the bug

Im trying to type my `uischema` object using the specified type in the uischema prop `UISchemaElement`
![Screenshot 2023-03-07 at 22 26 05](https://user-images.githubusercontent.com/32658690/223619719-2bb054b7-efd4-4d33-8c5c-72f57f375ad1.png)

### Expected behavior

I expect the object to be typed with the appropriate values and specs of the UI schema object

### Steps to reproduce the issue

I think the object type is incorrectly typed. and warns me with the error in the screenshot.

Here is a [conde sandbox](https://codesandbox.io/p/sandbox/json-forms-bug-2d7qpl?file=%2Fsrc%2FApp.tsx) with the error

### Screenshots

![Screenshot 2023-03-07 at 22 26 32](https://user-images.githubusercontent.com/32658690/223619820-d3bad544-ec32-4eab-81bc-a6c52a2e482f.png)

### In which browser are you experiencing the issue?

No browser. this is a type error

### Which Version of JSON Forms are you using?

v3.0.0

### Framework

React

**Comment by sdirix** [comment]
Thanks for the report!

Yes, our types are not set up in a convenient way so that Typescript is not able to automatically determine that your uischema is a valid `UISchemaElement`.

In this case the `uischema` needs to be typed as `VerticalLayout` and the child element as `ControlElement`, then the type errors will be gone.

**Comment by dcantu96** [comment]
Thank you. I did what you said and the errors are for the first `elements` are gone however, there are still some errors in the children elements of the uiSchema. I just copied the example in your home page https://jsonforms.io/

![Screenshot 2023-03-09 at 8 28 01](https://user-images.githubusercontent.com/32658690/224055167-eec9482f-2ad7-4cda-9b15-511870d18ac6.png)

**Comment by dcantu96** [comment]
I will just type it as any for now, just letting you know. Apart from this thanks for this library it's awesome ❤️

**Comment by sdirix** [comment]
Sadly each element needs to be typed for Typescript to recognize that they are compatible, e.g.

```ts
const uischema: VerticalLayout = {
  type: 'VerticalLayout',
  elements: [
    {
      type: 'Control',
      scope: '#/properties/firstName'
    } as ControlElement,
  ],
};
```

To avoid the forced cast a separate variable is needed, e.g.

```ts
const firstNameControl: ControlElemenet = {
  type: 'Control',
  scope: '#/properties/firstName'
}
const uischema: VerticalLayout = {
  type: 'VerticalLayout',
  elements: [ firstNameControl ],
};
```

Note that this behavior is not intended, we just set up the types in an unfortunate way. It's definitely on our list of things to improve.

**Comment by neelav** [comment]
Is there any update on when this would be fixed?

**Comment by sdirix** [comment]
As you can see with the milestone planning we eventually want to tackle this but there is no specific date which I could announce. Of course, if you, or someone else, would like to contribute, then this could be brought in much earlier.

**Comment by MartinDavi** [comment]
Hey as someone evaluating JsonForms for a large complex project I will say that hitting typescript issues immediately is a pretty much a non-starter at this point.

Understanding of course this is open source, free, etc! The project looks very nice but if the types aren't correct internally, that's gonna be a deal breaker for many I imagine.

**Comment by sdirix** [comment]
The types are complete, however they are in a state which is not that convenient to use. They are set up in hierarchical fashion, i.e. `VerticalLayout extends UiSchemaElement`. Therefore if you have a proper `VerticalLayout` typed element in hand, then it will be accepted by all functions accepting a `UiSchemaElement`. However if it's just an implicit type, then Typescript fails to realize that it's valid. Therefore you need to manually assert the children types.

**Comment by MartinDavi** [comment]
Gotchya - yes my surprise is that `uischema` prop is not typed more explicitly and instead is relying on typescript allowing extra properties not defined in the interface to be passed in.

Anyway, all good - I'll keep trucking along. Thanks for the response!

**Comment by IceFreez3r** [comment]
Would it be possible to fix this, by using the current `UiSchemaElement` as the base element (`UiSchemaBaseElement`) and then define `uischema` as `AnyUiSchemaElement` as a union of all complex types? If yes, can you estimate how hard it would be to swap that? I would be willing to invest a few hours to contribute a fix.

**Comment by sdirix** [comment]
Yes, some sort of union type would work. We could then refer to the union type wherever we currently refer to a `UiSchemaElement`.

Ideally we should also support custom elements, because adopters can write arbitrary custom renderers, consuming arbitrary ui schemas. However I don't like to define a generic ` [key: string]: unknown` type, as this would allow to hand over everything, removing all the convenience of type checking. So instead we should probably accept type attributes for some parts of our API, where the adopter can hand in their own type instead. However this could be a follow up.

@lucas-koehler What do you think of such a union type approach? Any better/other ideas?

**Comment by lucas-koehler** [comment]
Hi @sdirix @IceFreez3r ,

I agree with the union type idea. I think this is a good improvement to the typing.
From a type naming perspective I think it would be clearest to use `UISchemaElement` as the type specifying the union.
I.e.
`type UISchemaElement = HorizontalLayout | VerticalLayout | GroupLayout | ControlElement | ...`
Introduce a new type `BaseUISchemaElement` that is used as the new base type to be extended by the concrete types. Meaning the properties of the current `UISchemaElement` are moved to `BaseUISchemaElement`.

For backwards compatibility reasons we might also want to include `BaseUISchemaElement` in the union.

@sdirix Do you agree with this proposal?

> So instead we should probably accept type generics for some parts of our API, where the adopter can hand in their own type instead. However this could be a follow up.

This would certainly be a nice improvement. I am in favor of doing this as a follow up in any case to make each of the changes easier to implement and understand.

It is hard for me to estimate how much effort this exactly is to implement but I'd hope that adapting the `UISchemaElement` type to the union does not lead to required changes in most places of the code.

**Comment by sdirix** [comment]
Yes, I think that's a good idea.

@IceFreez3r If you like, then please contribute the change as refined by @lucas-koehler. If you run into issues, feel free to open a draft PR so we can discuss the issue there.
